rust/src/dcerpc/dcerpc.rs

```diff
@@ -945,77 +903,40 @@ impl DCERPCState {
             }
         }
 
-        // Overwrite the dcerpc_state data in case of multiple complete queries in the
-        // same direction
-        if self.prev_dir == direction {
-            self.data_needed_for_dir = direction;
-        }
-
-        let buffer = match direction {
-            Direction::ToServer => {
-                if self.buffer_ts.len() + input_len > 1024 * 1024 {
-                    SCLogDebug!("DCERPC TOSERVER stream: Buffer Overflow");
-                    return AppLayerResult::err();
-                }
-                v = self.buffer_ts.split_off(0);
-                v.extend_from_slice(cur_i);
-                v.as_slice()
-            }
-            Direction::ToClient => {
-                if self.buffer_tc.len() + input_len > 1024 * 1024 {
-                    SCLogDebug!("DCERPC TOCLIENT stream: Buffer Overflow");
-                    return AppLayerResult::err();
-                }
-                v = self.buffer_tc.split_off(0);
-                v.extend_from_slice(cur_i);
-                v.as_slice()
-            }
-        };
-
-        if self.data_needed_for_dir != direction && !buffer.is_empty() {
-            return AppLayerResult::err();
-        }
-
-        // Set data_needed_for_dir in the same direction in case there is an issue with upcoming parsing
-        self.data_needed_for_dir = direction;
-
+        let mut frag_bytes_consumed: u16 = 0;
         // Check if header data was complete. In case of EoF or incomplete data, wait for more
         // data else return error
-        if self.bytes_consumed < DCERPC_HDR_LEN.into() && input_len > 0 {
-            parsed = self.process_header(buffer);
+        if self.header.is_none() && !cur_i.is_empty() {
```

Comment:
With the direction based buffers gone, it's incorrect to use just one header per state. This should move in the transaction and the logic here should be reworked.

Reply:
do you have a tx while parsing the header?

Reply:
Not necessarily. If the request and response came in the expected order then I should have a tx while parsing the response header.. Trying to figure out what's the right way to handle all edge cases.

Reply:
Can we even have multiple partial headers in flight for the same direction? If not, I would just put this in the state instead of the tx.

Reply:
If parsing is done, the header could move into the tx if needed.

Reply:
Created a new rev assuming that the answer to 

> Can we even have multiple partial headers in flight for the same direction?

is no.